src/main/java/org/apache/sysml/runtime/instructions/cp/ParamservBuiltinCPInstruction.java

```diff
@@ -19,23 +19,264 @@
+				workerNum = Math.min(workerNum, vcores);
+			}
+			break;
+		case REMOTE_SPARK:
+			throw new DMLRuntimeException("Do not support remote spark.");
+		}
+		return workerNum;
+	}
+
+	/**
+	 * Create a server which serves the local or remote workers
+	 *
+	 * @return parameter server
+	 */
+	private ParamServer createPS() {
+		ParamServer ps = null;
+		switch (_mode) {
+		case LOCAL:
+			ps = new LocalParamServer();
+			break;
+		case REMOTE_SPARK:
+			throw new DMLRuntimeException("Do not support remote spark.");
+		}
+		return ps;
+	}
+
+	private long getBatchSize() {
+		if (!getParameterMap().containsKey(PS_BATCH_SIZE)) {
+			return DEFAULT_BATCH_SIZE;
+		}
+		long batchSize = Integer.valueOf(getParam(PS_BATCH_SIZE));
+		if (batchSize <= 0) {
+			throw new DMLRuntimeException(String.format(
+					"Paramserv function: the number of argument '%s' could not be less than or equal to 0.",
+					PS_BATCH_SIZE));
+		}
+		return batchSize;
+	}
+
+	private ListObject getHyperParams(ExecutionContext ec) {
+		ListObject hyperparams = new ListObject(Collections.emptyList());
+		if (getParameterMap().containsKey(PS_HYPER_PARAMS)) {
+			hyperparams = ec.getListObject(getParam(PS_HYPER_PARAMS));
+		}
+		return hyperparams;
+	}
+
+	private void doDataPartition(ExecutionContext ec, List<LocalPSWorker> workers) {
+		MatrixObject features = ec.getMatrixObject(getParam(PS_FEATURES));
+		MatrixObject labels = ec.getMatrixObject(getParam(PS_LABELS));
+		MatrixObject valFeatures = ec.getMatrixObject(getParam(PS_VAL_FEATURES));
+		MatrixObject valLabels = ec.getMatrixObject(getParam(PS_VAL_LABELS));
+		PSScheme scheme = DEFAULT_SCHEME;
+		if (getParameterMap().containsKey(PS_SCHEME)) {
+			scheme = PSScheme.valueOf(getParam(PS_SCHEME));
+		}
+		switch (scheme) {
+		case DISJOINT_CONTIGUOUS:
+			disjointContiguous(features, labels, valFeatures, valLabels, workers);
+			break;
+		case DISJOINT_RANDOM:
+		case OVERLAP_RESHUFFLE:
+		case DISJOINT_ROUND_ROBIN:
+			throw new DMLRuntimeException(
+					String.format("Paramserv function: the scheme '%s' is not supported.", scheme));
+		}
+	}
+
+	private void disjointContiguous(MatrixObject features, MatrixObject labels, MatrixObject valFeatures,
+			MatrixObject valLabels, List<LocalPSWorker> workers) {
+		// training data
+		List<MatrixObject> pfs = disjointContiguous(features);
+		List<MatrixObject> pls = disjointContiguous(labels);
+		for (int i = 0; i < workers.size(); i++) {
+			workers.get(i).setFeatures(pfs.get(i));
+			workers.get(i).setLabels(pls.get(i));
+		}
+
+		// validation data
+		List<MatrixObject> pvfs = disjointContiguous(valFeatures);
+		List<MatrixObject> pvls = disjointContiguous(valLabels);
+		for (int i = 0; i < workers.size(); i++) {
+			workers.get(i).setValFeatures(pvfs.get(i));
+			workers.get(i).setValLabels(pvls.get(i));
+		}
+	}
+
+	private List<MatrixObject> disjointContiguous(MatrixObject mo) {
+		List<MatrixObject> list = new ArrayList<>();
+		long stepSize = mo.getNumRows() / _workerNum;
```

Comment:
Please use a Math.ceil - otherwise you would get more partitions than workers or miss the last batch.